### PR TITLE
Throw clear errors messages for model checking index errors

### LINKED ISSUE #11: Model checking failures - Bug
The `ModelChecker.is_true()` method fails when evaluating 0-ary predicates (predicates with no arguments) that have numpy array extensions, raising a `ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()`.

### Error Message

```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Traceback:**
```python
File "/home/staff/a/akhedekar/chemlog-peptides/chemlog/fol_classification/model_checking.py", line 328, in find_model
    literals = [
File "/home/staff/a/akhedekar/chemlog-peptides/chemlog/fol_classification/model_checking.py", line 331, in <listcomp>
    if len(get_vars_in_formula(literal).intersection(variables)) > 0 or self.is_true(literal)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

### Minimal Reproducible example

```python
    def test_model_checking(self, reasoner: GavelFOLReasoner):
        """Test that exceptions in does_mol_match_tptp_definition are properly raised."""
        formula_str = "cation <=> net_charge_positive"

        _, parsed_formula = reasoner.get_tptp_fol_definition(formula_str)

        mol = Chem.MolFromSmiles(
            "C(=O)(C1=CC=C(C=C1F)OCCCCCC[NH+](CC=C)C)C=2C=CC(=CC2)Br"
        )

        reasoner.does_mol_match_tptp_definition(mol, parsed_formula)

```

### Expected Behavior

The model checker should successfully evaluate 0-ary predicates and return a boolean value (True/False) that can be used in boolean contexts.

### Actual Behavior

When `is_true()` returns a numpy array for a 0-ary predicate at line 164:

https://github.com/sfluegel05/chemlog-peptides/blob/7f4102e103452f0969b673700c476bc6e082b98a/chemlog/fol_classification/model_checking.py#L164

This array is then returned at line 206 and used in a boolean context (`or` operator at line 331), causing numpy to raise a `ValueError` because it cannot determine if a multi-element array should be considered True or False.
https://github.com/sfluegel05/chemlog-peptides/blob/7f4102e103452f0969b673700c476bc6e082b98a/chemlog/fol_classification/model_checking.py#L331

### Root Cause

In model_checking.py, the `is_true()` method at lines 159-164:

```python
if len(literal.arguments) > 1:
    res = self.extensions[literal.predicate][tuple(literal.arguments)]
elif len(literal.arguments) == 1:
    res = self.extensions[literal.predicate][literal.arguments[0]]
else:
    res = self.extensions[literal.predicate]  # Line 164 - problematic for numpy arrays
```

For 0-ary predicates (no arguments), the code directly returns `self.extensions[literal.predicate]` without extracting the scalar value from the numpy array. When this result is used in a boolean context (e.g., `if ... or res`), numpy raises the ambiguity error.

cc: @sfluegel05

### PR DESCRIPTION
**Pull request by aditya0by0**
- Closes #11 

Uses this changes in https://github.com/ChEB-AI/chebai-NL2FOL/pull/11

### CODE REVIEW

**Comment by aditya0by0** [comment]
@sfluegel05, Please refer to the test cases. I have added these test cases for all known errors till now (https://github.com/sfluegel05/chemlog-peptides/issues/11 and https://github.com/sfluegel05/chemlog-peptides/pull/9).

Copilot has modified `model_checker.py` file to throw clear error message instead of index errors. 
The model checking core logic is untouched but additional code is generated to enable to to throw clear errors. As the model checking file is modified, hence, need your **intricate review for this PR** _to avoid any disruption of existing functionality_.

**Comment by aditya0by0** [comment]
Also, @sfluegel05  could you please add me as collaborator to the peptides repo so that I can raise PR directly in your repo and this will also enable me to ask copilot to review the PR. Right now this is not possible

File: chemlog/fol_classification/model_checking.py [hunk 1]
```diff
@@ -127,6 +131,182 @@ def __init__(
...
+            extension = self.extensions[predicate_name]
+            if isinstance(extension, np.ndarray):
+                return extension.ndim
+            return 0
+        if predicate_name in self.definitions:
+            return len(self.definitions[predicate_name][0])
+        return None
+
+    @staticmethod
+    def _variable_name(variable: logic.Variable) -> str:
```
Comment (sfluegel05) [inline chemlog/fol_classification/model_checking.py]: I don't see why this method is needed. Variables always have a symbol and `str(variable.symbol)` should be the same as `str(variable)`

**Review by sfluegel05 — COMMENTED** [review]
I don't have the time to go through all 200+ lines of added code right now. It does not look like it interferes with the model checking. Probably this code could be more concise but if it works it works

File: chemlog/fol_classification/model_checking.py [hunk 2]
```diff
@@ -161,33 +341,36 @@ def is_true(self, literal: logic.LogicExpression) -> bool:
...
                         f" {len(literal.arguments)} arguments"
                     )
+                validated_arguments = self._validate_literal_arguments(literal)
 
                 if np.isnan(
                         self.calculated_extensions[literal.predicate][
-                            tuple(literal.arguments)
+                            validated_arguments
                         ]
                 ):
```
[the remark this answers is not in the document]
Comment (aditya0by0) [reply chemlog/fol_classification/model_checking.py:368]: @copilot, so a fix is needed for this or not. Can you elaborate?

File: chemlog/fol_classification/model_checking.py [hunk 3]
```diff
@@ -161,33 +341,36 @@ def is_true(self, literal: logic.LogicExpression) -> bool:
...
                     )
+                validated_arguments = self._validate_literal_arguments(literal)
                 if len(literal.arguments) > 1:
-                    res = self.extensions[literal.predicate][tuple(literal.arguments)]
+                    res = self.extensions[literal.predicate][validated_arguments]
                 elif len(literal.arguments) == 1:
-                    res = self.extensions[literal.predicate][literal.arguments[0]]
+                    res = self.extensions[literal.predicate][validated_arguments[0]]
                 else:
                     res = self.extensions[literal.predicate]
```
[the remark this answers is not in the document]
Comment (aditya0by0) [reply chemlog/fol_classification/model_checking.py:353]: Not sure what this exactly means,  @sfluegel05 please let me know if you have any idea about it.

**Comment by aditya0by0** [comment]
> I don't have the time to go through all 200+ lines of added code right now. It does not look like it interferes with the model checking. Probably this code could be more concise but if it works it works

@sfluegel05, Thanks for the review. True that right now its not possible to review all the code.  I have also added multiple test cases to cover edge cases as well as the usual functionality. In particular, there are test cases that should pass without errors during model checking. Based on my current/recent understanding of ChemLog, I included the pass cases I could identify.

If you think there are additional passing test cases that would be valuable for model checking, please feel free to add or suggest them.

**Comment by aditya0by0** [comment]
@sfluegel05, Also please let me know whether this can merged 

OR

we hold on this PR,  and use this chemlog branch for learning.

**Comment by sfluegel05** [comment]
> we hold on this PR, and use this chemlog branch for learning.

let's do the learning from this branch. If it works, you can merge it to the dev branch

**Comment by aditya0by0** [comment]
> > we hold on this PR, and use this chemlog branch for learning.
> 
> let's do the learning from this branch. If it works, you can merge it to the dev branch

Here, an example which shows the logic works.  (See last message in conversation history)
https://smith.langchain.com/public/53334ea9-b4c6-4bfd-8dcb-9a686687e307/r
https://smith.langchain.com/public/eb8a8d91-dd1e-496b-b1c3-d14e4f70bbbf/r

**Comment by aditya0by0** [comment]
> > we hold on this PR, and use this chemlog branch for learning.
> 
> let's do the learning from this branch. If it works, you can merge it to the dev branch

@sfluegel05 
Its working and learning is completed. As suggested I will proceed to merge this.
